cloud/testserver/deployments.go

```diff
@@ -7,160 +7,151 @@ import (
-	nextStackID int64
-	// as hacky as it can get:
-	// map of organization -> (map of deployment_id -> (map of stack_id -> deployment))
-	deployments map[string]map[string]map[int64]cloud.DeploymentStackRequest
+// GetDeployments is the GET /deployments handler.
+func GetDeployments(store *cloudstore.Data, w http.ResponseWriter, _ *http.Request, p httprouter.Params) {
+	orguuid := cloud.UUID(p.ByName("orguuid"))
+	deployuuid := cloud.UUID(p.ByName("deployuuid"))
 
-	events map[string]map[string]map[string][]string
-}
+	org, found := store.GetOrg(orguuid)
+	if !found {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
 
-func newDeploymentEndpoint() *deploymentHandler {
-	return &deploymentHandler{
-		deployments: make(map[string]map[string]map[int64]cloud.DeploymentStackRequest),
-		events:      make(map[string]map[string]map[string][]string),
+	deploymentInfo, found := store.GetDeployment(&org, deployuuid)
+	if !found {
+		w.WriteHeader(http.StatusNotFound)
+		return
 	}
+	w.Header().Add("Content-Type", "application/json")
+	marshalWrite(w, deploymentInfo)
 }
 
-func (dhandler *deploymentHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	dhandler.mu.Lock()
-	defer dhandler.mu.Unlock()
-
-	params := httprouter.ParamsFromContext(r.Context())
-	orguuid := params.ByName("orguuid")
-	deployuuid := params.ByName("deployuuid")
-
-	if !strings.HasPrefix(r.Header.Get("User-Agent"), "terramate/") {
-		w.WriteHeader(http.StatusBadRequest)
+// CreateDeployment is the POST /deployments handler.
+func CreateDeployment(store *cloudstore.Data, w http.ResponseWriter, r *http.Request, p httprouter.Params) {
+	defer func() { _ = r.Body.Close() }()
+	data, _ := io.ReadAll(r.Body)
+	var rPayload cloud.DeploymentStacksPayloadRequest
+	err := json.Unmarshal(data, &rPayload)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		writeErr(w, errors.E(err, "failed to unmarshal data: %s", data))
 		return
 	}
 
-	if dhandler.deployments[orguuid] == nil {
-		dhandler.deployments[orguuid] = make(map[string]map[int64]cloud.DeploymentStackRequest)
-		dhandler.events[orguuid] = make(map[string]map[string][]string)
-	}
-	if dhandler.deployments[orguuid][deployuuid] == nil {
-		dhandler.deployments[orguuid][deployuuid] = make(map[int64]cloud.DeploymentStackRequest)
-		dhandler.events[orguuid][deployuuid] = make(map[string][]string)
+	err = rPayload.Validate()
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		writeErr(w, err)
+		return
 	}
 
-	w.Header().Add("Content-Type", "application/json")
+	stackCommands := map[string]string{}
 
-	if strings.HasSuffix(r.URL.Path, "/events") {
-		if r.Method != "GET" {
-			w.WriteHeader(http.StatusMethodNotAllowed)
+	// deployment commit_sha is not required but must be present in all test cases.
+	// TODO(i4k): review this!!!
+	for _, st := range rPayload.Stacks {
+		if st.CommitSHA == "" {
+			w.WriteHeader(http.StatusBadRequest)
+			_, _ = w.Write([]byte(`commit_sha is missing`))
```

Comment:
It would be nice to handle the error here in case this ever fails, otherwise it would make the failing test hard to debug.